### PR TITLE
JSON export and rehydration (issue 30)

### PR DESCRIPTION
**Pull request by qpwo**
Will add rehydration later

### CODE REVIEW

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt [hunk 1]
```diff
@@ -150,11 +151,12 @@ fun main(argv: Array<String>) {
...
-                    val jsonBytes = it.getInputStream("$zipEntryName.json")?.readAllBytes()
-
+                    val jsonBytes = it.getInputStream("$zipEntryName.json")?.readBytes()
                     if (jsonBytes != null) {
-                        val outFileName = documentOutputDir.resolve("${document.name}-${annotationRecord.user}.json");
-                        Files.write(outFileName, jsonBytes)
+                        var jsonTree = ObjectMapper().readTree(jsonBytes) as ObjectNode
+                        jsonTree.replaceFieldEverywhere("sofaString", "__DOCUMENT_TEXT_REDACTED_FOR_IP_REASONS__")
+                        val outFileName = documentOutputDir.resolve("${document.name}-${annotationRecord.user}.json")
+                        Files.write(outFileName, jsonTree.toString().toByteArray())
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: In general, `toString` is to be avoided except for diagnostic purposes. Prefer to generate the JSON string using `ObjectMapper`.

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt [hunk 2]
```diff
@@ -150,11 +151,12 @@ fun main(argv: Array<String>) {
...
                     }
 
-                    val jsonBytes = it.getInputStream("$zipEntryName.json")?.readAllBytes()
-
+                    val jsonBytes = it.getInputStream("$zipEntryName.json")?.readBytes()
                     if (jsonBytes != null) {
-                        val outFileName = documentOutputDir.resolve("${document.name}-${annotationRecord.user}.json");
-                        Files.write(outFileName, jsonBytes)
+                        var jsonTree = ObjectMapper().readTree(jsonBytes) as ObjectNode
+                        jsonTree.replaceFieldEverywhere("sofaString", "__DOCUMENT_TEXT_REDACTED_FOR_IP_REASONS__")
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: Add a comment about why we do this (that we cannot redistribute LDC documents and the user can always restore from those LDC documents)

File: pom.xml [hunk 3]
```diff
@@ -63,7 +63,7 @@
 			<dependency>
 				<groupId>edu.isi.vista</groupId>
 				<artifactId>common-core-open</artifactId>
-				<version>8.2.1-SNAPSHOT</version>
+				<version>8.2.2-SNAPSHOT</version>
```
Comment (rgabbard) [inline pom.xml]: What are we depending on which is not present in the release version?

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt [hunk 4]
```diff
@@ -100,38 +125,27 @@ fun main(argv: Array<String>) {
                     .resultObjectThrowingExceptionOnFailure<AeroResult<AnnotatorRecord>>(mapper)
                     .body
 
-            val documentOutputDir = projectOutputDir.resolve(document.name)
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: Why was this removed?

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt [hunk 5]
```diff
@@ -172,21 +190,37 @@ private data class Document(val id: Long, val name: String, val state: String) {
...
+    if (maxTries == 0)
+        throw RuntimeException("HTTP request has failed. Aborting.")
+    return when (result) {
+        is Result.Success<*> -> result.get()
+        is Result.Failure<*> -> {
+            logger.warn {
+                "Request timed out. Waiting $timeoutInSeconds seconds and trying again."
+            }
+            TimeUnit.SECONDS.sleep(timeoutInSeconds)
+            this.retryOnFailure(maxTries - 1, timeoutInSeconds)
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: In this case, I think the simple for loop is clearer than recursion

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt [hunk 6]
```diff
@@ -172,21 +190,37 @@ private data class Document(val id: Long, val name: String, val state: String) {
...
             result.get()
         }
     }
-}
\ No newline at end of file
+}
+
+private fun Request.retryOnFailure(maxTries: Int = 3, timeoutInSeconds: Long = 30): ByteArray {
+    val (_, _, result) = this.response()
+    if (maxTries == 0)
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: Always use braces

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import edu.isi.nlp.io.OffsetIndexedCorpus
+import edu.isi.nlp.symbols.Symbol
+import edu.isi.nlp.io.DocIDToFileMappings
+import com.google.common.base.Optional
+import com.fasterxml.jackson.databind.ObjectMapper
+import edu.isi.nlp.parameters.Parameters
+import java.io.File
+
+/**
+ * Puts the document text back into the 'sofaString' field inside json files created by `ExportAnnotations.kt`
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: Explain where and why they were removed

Comment (rgabbard) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: Also mention this program in the repo's `README.md`

Comment (rgabbard) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: Also, while we use the "dehydrate/rehydrate" language informally, it would be clearer to future readers if we called this something like `RestoreDocumentText`

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+fun main(argv: Array<String>) {
+    // Load parameters:
+    val params = Parameters.loadSerifStyle(File(argv[0]))
+    val indexDirectory = params.getExistingDirectory("indexDirectory")
+    val gigawordDataDirectory = params.getExistingDirectory("gigawordDataDirectory")
+    val jsonInputDirectory = params.getCreatableDirectory("dehydratedJsonDirectory")
+    val outputDirectory = params.getCreatableDirectory("rehydratedJsonDirectory")
+    outputDirectory.toString()
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: ?

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+ *     <li> {@code indexDirectory} is the location of the files produced by {@code IndexFlatGigaword.java}
+            (https://github.com/isi-vista/nlp-util/blob/master/nlp-core-open/src/main/java/edu/isi/nlp/corpora/gigaword/IndexFlatGigaword.java) </li>
+ *     <li> {@code gigawordDataDirectory} is the location of the gigaword text files </li>
+ *     <li> {@code jsonInputDirectory} is the location of the 'dehydrated' json files produced by {@code ExportAnnotations.kt} </li>
+ *     <li> {@code rehydratedJsonDirectory} is where the new json files will go </li>
+ *  </ul>
+ */
+
+fun main(argv: Array<String>) {
+    // Load parameters:
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: This comment doesn't add much :-)

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    val params = Parameters.loadSerifStyle(File(argv[0]))
+    val indexDirectory = params.getExistingDirectory("indexDirectory")
+    val gigawordDataDirectory = params.getExistingDirectory("gigawordDataDirectory")
+    val jsonInputDirectory = params.getCreatableDirectory("dehydratedJsonDirectory")
+    val outputDirectory = params.getCreatableDirectory("rehydratedJsonDirectory")
+    outputDirectory.toString()
+
+    // Create an OriginalTextSource for getting original document text to put in json:
+    val textMap = DocIDToFileMappings.forFunction { symbol ->
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: You could consider pulling everything from here through lines 44 into its own method (possibly just taking the `Parameters` as an argument)

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+            continue
+        }
+        val projectOutDir = File(outputDirectory, projectDir.name)
+        projectOutDir.mkdir()
+        for (jsonFile in projectDir.listFiles()) {
+            val jsonTree = objectMapper.readTree(jsonFile)
+            // First 21 characters of the filename are the document id
+            // Example filename is AFP_ENG_19960918.0012-admin.json
+            val docID = Symbol.from(jsonFile.name.substring(0, 21))
+            val text = textSource.getOriginalText(docID).get()
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: Check and raise an exception when `getOriginalText` returns `Optional.absent`. A possible idiom for this is:

```
val text = textSource.getOriginalText(docId).orNull() ?: 
       throw RuntimeException("Could not get original text for $docId")
```

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+            jsonTree.replaceFieldEverywhere("sofaString", text)
+            val outFile = File(projectOutDir, jsonFile.name)
+            outFile.writeBytes(prettyPrinter.writeValueAsBytes(jsonTree))
+        }
+    }
+}
+
+fun docIDToFilename(symbol: Symbol?): String? {
```
Comment (rgabbard) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: Why do we accept `Symbol?` here rather than requiring `Symbol`?

File: annotation-utils/pom.xml [hunk 13]
```diff
@@ -93,6 +93,11 @@
 			<groupId>edu.isi.vista</groupId>
 			<artifactId>common-core-open</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>edu.isi.vista</groupId>
+			<artifactId>nlp-core-open</artifactId>
+            <version>8.2.2-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (qpwo) [reply annotation-utils/pom.xml]: Deferring to @berquist here

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt (hunk 12 above)
Comment (qpwo) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: `DocIDToFileMappings.forFunction` sends a `Symbol?` but I made `docIDToFilename` throw an error

File: annotation-utils/pom.xml (hunk 13 above)
[the remark this answers is not in the document]
Comment (berquist) [reply annotation-utils/pom.xml]: I think this was done in the struggle to get things compiling, where the crucial step was adding the dependency to the inner pom, not specifically making it be `-SNAPSHOT`. @qpwo Can you try making all of the `-SNAPSHOT`s into the previous stable version and see if it still works?

**Comment by berquist** [comment]
I think my commit https://github.com/isi-vista/curated-training-annotator/pull/33/commits/16bf9e11325879e5402f0552eaa31875cdca66b3 can be reverted (restore the deleted retry function and delete the recursive extension method) since we discovered the problem is with a particular annotation account on the server.

File: pom.xml [hunk 14]
```diff
@@ -63,12 +63,12 @@
...
 			</dependency>
 			<dependency>
 				<groupId>edu.isi.vista</groupId>
 				<artifactId>nlp-core-open</artifactId>
-				<version>8.1.0</version>
+				<version>8.2.2-SNAPSHOT</version>
```
Comment (berquist) [inline pom.xml]: Try depending on `8.2.1`.

File: pom.xml [hunk 15]
```diff
@@ -63,12 +63,12 @@
 			<dependency>
 				<groupId>edu.isi.vista</groupId>
 				<artifactId>common-core-open</artifactId>
-				<version>8.2.1-SNAPSHOT</version>
+				<version>8.2.2-SNAPSHOT</version>
```
Comment (berquist) [inline pom.xml]: Try depending on `8.2.1`.

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+            val outFile = File(projectOutDir, jsonFile.name)
+            outFile.writeBytes(prettyPrinter.writeValueAsBytes(jsonTree))
+        }
+    }
+}
+
+fun makeTextSource(params: Parameters): OriginalTextSource {
+    // Create an OriginalTextSource for getting original document text to put in json:
+    val textMap = DocIDToFileMappings.forFunction { symbol ->
+        Optional.of(File(params.getExistingDirectory("gigawordDataDirectory"), docIDToFilename(symbol)))
```
Comment (berquist) [inline annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: Why is this `Optional`?

Comment (qpwo) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: `DocIDToFileMappings.forFunction` wouldn't take it non-optionally and I couldn't get an `as`-cast to work either

Comment (rgabbard) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: It expects a function which returns `Optional` because in general "I can't map this" is a valid response from a mapping strategy.

File: pom.xml (hunk 15 above)
Comment (qpwo) [reply pom.xml]: Couldn't get it to work with 8.2.2. IntelliJ couldn't find it

Comment (rgabbard) [reply pom.xml]: Does it work for 8.2.1?

Comment (rgabbard) [reply pom.xml]: It looks like changes since 8.2.1 pertain only to `IndexFlatGigaword`: https://github.com/isi-vista/nlp-util/commits/master

Comment (qpwo) [reply pom.xml]: Oops! I couldn't get it to work for 8.2.**1** but it works for 8.2.**2** so I think we should keep these changes.

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt (hunk 4 above)
Comment (qpwo) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: The output structure had one directory per file so we flattened it into just one directory per project. Should we revert that?

Comment (rgabbard) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: How do we handle multiple annotators annotating the same file on the same project?

Comment (qpwo) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: Oh is that possible? I thought each user has their own fresh copy of the project? (E.g. named `ArtifactExistence.DamageDestroy-bob_loblaw`)

**Comment by qpwo** [comment]
Did some more testing and indeed `8.2.2-SNAPSHOT` is required

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt (hunk 4 above)
Comment (rgabbard) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: By policy, yes, but it is possible in the Inception data model for the someone else to come in an annotate on the project with someone else's name on it, so we should handle that case.

Comment (qpwo) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: The containing dir names the project owner, and the file itself names the annotator. Does that eliminate possibility of collision? For example there is `output/ArtifactExistence.DamageDestroy-gabbard/AFP_ENG_19960918.0012-admin.json`

Comment (rgabbard) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: Yep, that will do the trick. Maybe add a comment somewhere to clarify this?

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt [hunk 17]
```diff
@@ -58,7 +58,9 @@ fun main(argv: Array<String>) {
     }
     logger.info { "Connecting to Inception at $inceptionUrl" }
 
-    val mapper = ObjectMapper().registerKotlinModule()
+    val objectMapper = ObjectMapper()
```
[the remark this answers is not in the document]
Comment (rgabbard) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt]: Can you return this to its original form?  It is more idiomatic Jackson code; otherwise the reader is going to say "why did they do that?"

File: annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt (hunk 7 above)
Comment (rgabbard) [reply annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt]: Still need to rename away from re-hydrate

File: annotation-utils/pom.xml (hunk 13 above)
[the remark this answers is not in the document]
Comment (rgabbard) [reply annotation-utils/pom.xml]: 8.2.2 has now been released, so you should be able to use 8.2.2 (without `-SNAPSHOT`) as the value now

File: pom.xml (hunk 15 above)
Comment (rgabbard) [reply pom.xml]: 8.2.2 here as well, and below

**Comment by qpwo** [comment]
I get this error when I use 8.2.2 without the SNAPSHOT and run `ExportAnnotations.kt`:
```
10:32:32.808 [main] WARN  mu.KLogging - com.github.kittinunf.fuel.core.FuelError: Cannot construct instance of `edu.isi.vista.annotationutils.Document` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: (BufferedInputStream); line: 1, column: 25] (through reference chain: edu.isi.vista.annotationutils.DocumentResult["body"]->java.util.ArrayList[0])
	com.github.kittinunf.fuel.core.FuelError$Companion.wrap(FuelError.kt:85)
	com.github.kittinunf.fuel.core.DeserializableKt.response(Deserializable.kt:176)
```

I think this statement is causing it:
```
1.        val documents = retryOnFuelError {
2.            getDocsUrl.httpGet(
3.                    parameters = listOf("projectId" to project.id))
4.                    .authenticateToInception()
5.                    .resultObjectThrowingExceptionOnFailure<DocumentResult>(mapper)
6.                    .body
7.        }
```

Where line 5 is the problem.

But why wouldn't this happen on the SNAPSHOT version? :thinking:

**Comment by rgabbard** [comment]
@qpwo : Are you sure the version of the `nlp-utils` dependencies reliably causes this difference?  If you look at  https://github.com/isi-vista/nlp-util/commits/master, you'll see that (unless you have local modifications to `nlp-utils`):

* `8.2.2` and `8.2.2-SNAPSHOT` are identical
* `8.2.1` and `8.2.2` differ only in code internal to `IndexFlatGigaword`, which should be irrelevant here
* `8.2.1` differs from `8.2.0` only in code internal to `Parameters`, which should be irrelevant here.\\

**Comment by qpwo** [comment]
You're right; I changed it back to `8.2.2-SNAPSHOT` and I got the same error as before. I changed it back to `8.2.2` and then did a `mvn clean install` and that killed the error somehow. Some sort of problem with changing dependencies in IntelliJ perhaps?

**Comment by rgabbard** [comment]
@qpwo : Had you edited the dependencies in IntelliJ's interface directly or only via the `pom.xml`?

**Comment by qpwo** [comment]
`pom.xml`

**Comment by rgabbard** [comment]
@qpwo : If the problem is not present on a clean build, let's not worry about it for now.

**Comment by qpwo** [comment]
Yes I agree I committed the changes earlier

**Comment by rgabbard** [comment]
@qpwo : Is this PR ready for re-review?

**Comment by qpwo** [comment]
Yes!

**Comment by berquist** [comment]
I'm now getting lots of `FuelError` messages, and things basically worked before except for the one problem with https://github.com/isi-vista/curated-training-annotator/issues/34. Something happened between https://github.com/isi-vista/curated-training-annotator/pull/33/commits/5ea83801d0345d3e173cba5a067370e550bedd3b and the PR head.

**Comment by rgabbard** [comment]
@qpwo : Poke me for re-review once you and @berquist have sorted out the problem above

**Comment by qpwo** [comment]
@berquist Are you still getting the error after a `mvn clean install` and then `invalidate caches and restart` in IntelliJ?

**Comment by qpwo** [comment]
Might be ready for a merge I'm still not getting errors
